### ISSUE TITLE
Smart Cards not working with the MITM component

### ISSUE THREAD

**Issue by keremenci** [GoSecure/pyrdp#382]
When connecting to a server via pyrdp MITM, smart cards are not working properly. The device appears on the remote host but the certificates in the smart card are unaccessible.

**Comment by obilodeau** [comment]
Supporting SmartCard is not something we are interested in at this point. I don't even know how to test it.

That said, if you are willing to do the exploration and the code changes required, we can guide you.

**Comment by keremenci** [comment]
I used EIDVirtual to create a smart card out of an usb stick. By installing EIDVirtual on the server, one can use the smart card manager to view the certificates on the smart card in the remote server.
It appears that the device is recognized but the certificates cannot be read from the card. I dumped the incoming and outgoing network packets but couldn't figure out the decryption in wireshark. Would love some help with that

**Comment by obilodeau** [comment]
Information about how to decrypt TLS traffic with Wireshark can be found in [Wireshark's wiki](https://wiki.wireshark.org/TLS). An easier-to-follow tutorial with screenshots is also available here: https://unit42.paloaltonetworks.com/wireshark-tutorial-decrypting-https-traffic/.

Record the whole conversation then open the pcap following the TLS decryption instructions. The TLS master-secrets will be in `pyrdp_output/logs/ssl.log`.

**Comment by keremenci** [comment]
After dumping the incoming and outgoing packets via tcpdump using the following filters:
* `sudo tcpdump -s 0 -i <interface> -w in.pcap src \<client ip> and dst port 3389`
* `sudo tcpdump -s 0 -i <interface> -w out.pcap src <mitm ip> and dst \<server ip> and dst port 3389`

I don't seem to be able to filter the traffic by rdp in wireshark, and all I see is a bit of TLS1.2 traffic followed by a large amount of TCP packets. I did configure wireshark to use the log field and also supplied the private key in the preferences > protocols > TLS

This is what i did in order:

* Run the first tcpdump command
* Run the second tcpdump
* Run the mitm
* Connect via rdp and try to read the smart card, fail
* Disconnect using the button in windows
* Stop the tcpdump processes
* Copy over the pcap files and the log file over to local
* Open it in wireshark

Any ideas what I may be missing? My networking knowledge is a bit limited so any help would be appreciated.

**Comment by keremenci** [comment]
Also, it seems that all the IO requests and responses pass through except the 0x000900a4 | FSCTL_SET_REPARSE_POINT call, which has something to do with SCardGetStatusChangeW (Couldn't really figure out this part).

**Comment by obilodeau** [comment]
> I did configure wireshark to use the log field and also supplied the private key in the preferences > protocols > TLS

Don't supply the private key. Supply the TLS master secrets. Load the `ssl.log` in the red square here: ![Wireshark screenshot for TLS config](https://github.com/GoSecure/pyrdp/raw/master/docs/screens/wireshark-tls.png)

**Comment by keremenci** [comment]
I tried various combination of supplying the pk and the log and was not successful

**Comment by keremenci** [comment]
![image](https://user-images.githubusercontent.com/57116822/149294138-9678907f-e52e-4485-8c45-d564590034f0.png)
These two IO requests (as defined in [the docs](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/30662c80-ec6e-4ed1-9004-2e6e367bb59f)) from the server never receive a response. Any ideas on what the cause may be?

**Comment by obilodeau** [comment]
I think it might be because we implemented MITM for Device IO since it is how we collect files but we haven't done anything regarding smart cards. One approach would be to use a debugger at the right spots and see if you could pass through the exchange back and forth instead of intercepting them.

For example:

install ipdb

```pip install ipdb
```

add a statement to trigger the debugger at the right spot. In `pyrdp/parser/rdp/virtual_channel/device_redirection.py`, under:
```
    def parseDeviceIORequest(self, stream: BytesIO) -> DeviceIORequestPDU:
```

Add:
```
import ipdb; ipdb.set_trace()
```

Restart `pyrdp-mitm` and see what is mishandled (it might not be the first time the debugger trips)

Good luck!

**Comment by keremenci** [comment]
It appears that the major function field is set to 0x0000000C and the minor function is 0 even though it is explicitly stated that this is not a valid state of the pdu in [the docs](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a087ffa8-d0d5-4874-ac7b-0494f63e2d5d). 
![image](https://user-images.githubusercontent.com/57116822/149524108-2ab38e86-0e90-4640-af62-07f1647734af.png)

**Comment by obilodeau** [comment]
You are not providing enough details. Without them, I can't help you.

* Is the MITM sending the 0, the real server or the real client?
* If its the MITM is it sending it to the client or to the server?
* Is the server query before the MITM correct or not?
* Where in the code is this value changed to a non-valid one?
* if you debug in `parseDirectoryControlResponse` from `pyrdp/parser/rdp/virtual_channel/device_redirection.py` does this help you?

**Comment by keremenci** [comment]
Yes, my bad. Had to tend to some other stuff and could not exactly pay attention. Sorry for any inconvenience this may have caused.

Now that i'm back, I realized that what I said here was incorrect. There is nothing wrong with the pdu.

Upon further investigation, and some parsing by hand, I managed to extract the ioctl codes for the smart card DeviceIOControl calls. I couldn't find anything about the codes that I parsed, but then stumbled across the definitions in the [smart card extension page](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpesc/0428ca28-b4dc-46a3-97c3-01887fa44a90)

There really should be no issue for smart cards, as all the pdu's that pass through are just forwarded with no interception, and I couldn't find anything that was mishandled.

Do we need to do some additional work to support the protocol extensions?

> It appears that the major function field is set to 0x0000000C and the minor function is 0 even though it is explicitly stated that this is not a valid state of the pdu in [the docs](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/a087ffa8-d0d5-4874-ac7b-0494f63e2d5d). ![image](https://user-images.githubusercontent.com/57116822/149524108-2ab38e86-0e90-4640-af62-07f1647734af.png)

**Comment by obilodeau** [comment]
Does that feature use DynamicChannels or VirtualChannels? We don't support DynamicChannels just yet. See #232 for the beginning of that work.

**Comment by Res260** [comment]
@olivier i checked, it's the rdpdr channel (device redirection)

**Comment by keremenci** [comment]
It seems that the server can access the reader while the issues manifest themselves when we try to access the contents of the card.

Certutil hangs, the EIDVirtual smart card manager hangs, SCM Microsystems PC/SC diagnostic tool identifies the device as working but trying to access it causes the process to hang
![image](https://user-images.githubusercontent.com/57116822/151775710-17555cf9-7955-4fb5-999b-98fbd79c4e63.png)

The behaviour is really chaotic, debugging drops the connection. Dumping the PDU information either causes the smart card manager of windows to not start at all, or drops the connection after the following exchange:

```
900e0 -> SCARD_IOCTL_ACCESSSTARTEDEVENT
90014 -> SCARD_IOCTL_ESTABLISHCONTEXT
9002c -> SCARD_IOCTL_LISTREADERSW
90108 -> SCARD_IOCTL_GETDEVICETYPEID
90014 -> SCARD_IOCTL_ESTABLISHCONTEXT
90108 -> SCARD_IOCTL_GETDEVICETYPEID
90018 -> SCARD_IOCTL_RELEASECONTEXT
900a4 -> SCARD_IOCTL_GETSTATUSCHANGEW
900a4 -> SCARD_IOCTL_GETSTATUSCHANGEW
90014 -> SCARD_IOCTL_ESTABLISHCONTEXT
9002c -> SCARD_IOCTL_LISTREADERSW
900a4 -> SCARD_IOCTL_GETSTATUSCHANGEW
90014 -> SCARD_IOCTL_ESTABLISHCONTEXT
90014 -> SCARD_IOCTL_ESTABLISHCONTEXT
900b0 -> SCARD_IOCTL_CONNECTW
900a4 -> SCARD_IOCTL_GETSTATUSCHANGEW
900bc -> SCARD_IOCTL_BEGINTRANSACTION
900a8 -> SCARD_IOCTL_CANCEL
900a8 -> SCARD_IOCTL_CANCEL
```

I am kind of lost at this moment. Any ideas?
